[docs] Extend channel surfing docs (#47288)
expo/expo: docs/pages/eas-update/channel-surfing.mdx

---
title: Channel surfing
description: Learn how to switch EAS Update channels at runtime.
---

import { BookOpen02Icon } from '@expo/styleguide-icons/outline/BookOpen02Icon';

import { BoxLink } from '~/ui/components/BoxLink';
import { Prerequisites, Requirement } from '~/ui/components/Prerequisites';
import { Terminal } from '~/ui/components/Snippet';

> Channel surfing lets an installed release build request updates from a different EAS Update channel at runtime.

Reviewer: ```suggestion
**Channel surfing** lets an installed release build request updates from a different EAS Update channel at runtime.
```

In the default EAS Update flow, a build includes an update URL and request headers at build time. The `expo-channel-name` request header determines which channel the build requests updates from. Channel surfing stores a runtime override for that request header on the device.

Channel surfing does not change native code or bypass update compatibility rules. The update on the selected channel must still match the installed app's platform and [runtime version](/eas-update/runtime-versions/).

> **warning** Channel surfing with `Updates.setUpdateRequestHeadersOverride()` is available in Expo SDK 54 with `expo-updates` version 0.29.0 and later.

## When to use channel surfing

Use channel surfing when the same native runtime can load updates from more than one channel.

> You can use channel surfing to:

Reviewer: ```suggestion
You can use channel surfing to:
```

> - Allow a single installed app to move between channels on demand. The app can switch channels at runtime instead of remaining tied to the channel defined at build time.

Reviewer: ```suggestion
- When the same native runtime can load updates from more than one channel.
- Allow a single installed app to move between channels on demand. The app can switch channels at runtime instead of remaining tied to the channel defined at build time.
```

It is confusing that the first line of this section instruct the reader on "when" to use it and then second line also has the same message with more items in a list.

> - Enable preview and testing on real builds so that developers, QA, or other stakeholders can try in-progress updates using the same production build that users have installed.

Reviewer: **[suggestion]** This sentence runs about 27 words. Splitting it lets each idea land on its own (`style-guide.md#core-principles`).

```suggestion
- Enable preview and testing on real builds. Developers, QA, and other stakeholders can try in-progress updates using the same production build that users have installed.
```

- Speed up iteration and validation by redirecting the app to another channel, without waiting for a new build.

> For developer-focused testing, use a [development build with `expo-dev-client`](/eas-update/expo-dev-client/) to preview compatible updates. This does not use `Updates.setUpdateRequestHeadersOverride()`, which is not supported in development builds.

Reviewer: ```suggestion
For developer-focused testing and to preview compatible updates, use a [development build with `expo-dev-client`](/eas-update/expo-dev-client/) library installed. This does not use `Updates.setUpdateRequestHeadersOverride()`, which is not supported in development builds.
```

  <Requirement title="A release build installed">
    Use a release build, or a debug build with
    [`EX_UPDATES_NATIVE_DEBUG`](/eas-update/debug/#debugging-of-native-code-while-loading-the-app-through-expo-updates)
    enabled. This is different from a development build with `expo-dev-client`. Most of the
    `expo-updates` API is unavailable in normal development builds.
  </Requirement>
  <Requirement title="A channel configured at build time">
    `Updates.setUpdateRequestHeadersOverride()` can only override request header keys that were
    embedded in the build. For channel surfing, this means the build must include
    `expo-channel-name`. [EAS Build](/build/introduction/) adds the channel from **eas.json** to
    your native project at build time. If you do not use EAS Build, configure the channel with
    [`updates.requestHeaders`](/eas-update/getting-started/#configure-update-channels-in-appjson) or
    in your native project.
  </Requirement>
  <Requirement title="A compatible update on the target channel">
    The target channel must have an update for the installed app's platform and runtime version.
  </Requirement>
</Prerequisites>

## Switch channels

> Provide an app-level trigger for changing channels, such as a hidden menu for trusted users or another mechanism that fits your workflow. When a channel is selected, use [`Updates.setUpdateRequestHeadersOverride()`](/versions/latest/sdk/updates/#updatessetupdaterequestheadersoverriderequestheaders) to override the `expo-channel-name` request header. After setting the override, check for an update, fetch it if available, and reload the app.

Reviewer: **[suggestion]** Introduce the code block with a colon, the way the `Terminal` example further down does ("...compatible channel:"). Right now this lead-in ends with a period.

```suggestion
Provide an app-level trigger for changing channels, such as a hidden menu for trusted users or another mechanism that fits your workflow. When a channel is selected, use [`Updates.setUpdateRequestHeadersOverride()`](/versions/latest/sdk/updates/#updatessetupdaterequestheadersoverriderequestheaders) to override the `expo-channel-name` request header. After setting the override, check for an update, fetch it if available, and reload the app:
```

> ```ts

Reviewer: ````suggestion
```tsx
````

import * as Updates from 'expo-updates';

export async function switchUpdateChannelAsync(channel: string) {
  Updates.setUpdateRequestHeadersOverride({
    'expo-channel-name': channel,
  });

  const update = await Updates.checkForUpdateAsync();

  if (update.isAvailable) {
    await Updates.fetchUpdateAsync();
  }

  await Updates.reloadAsync();
}
```

The override persists on the device. After it is set, future update checks use the selected channel until the app clears or replaces the override, or until the app is uninstalled.

## Switch back to the build channel

> Pass `null` to `Updates.setUpdateRequestHeadersOverride()` to clear the request header override and return to the channel configured in the build.

Reviewer: **[suggestion]** Same here: end the lead-in with a colon so it introduces the code block consistently with the other sections.

```suggestion
Pass `null` to `Updates.setUpdateRequestHeadersOverride()` to clear the request header override and return to the channel configured in the build:
```

> ```ts

Reviewer: ````suggestion
```tsx
````

import * as Updates from 'expo-updates';

export async function clearUpdateChannelOverrideAsync() {
  Updates.setUpdateRequestHeadersOverride(null);

  const update = await Updates.checkForUpdateAsync();

  if (update.isAvailable) {
    await Updates.fetchUpdateAsync();
  }

  await Updates.reloadAsync();
}
```

`Updates.channel` reflects the channel that was active when the app started. It does not update immediately after calling `Updates.setUpdateRequestHeadersOverride()`, but it reflects the new channel after the app reloads.

> ## Test channel surfing

Reviewer: It'd be nice if we can use `Step` component here and convert the following into steps. Will provide a clear structure and the instructions will be represented better (also gives less chance to skim for a human eye).

> To test channel surfing, install a release build that points to the channel you want to start from. If you use [EAS Build](/build/introduction/), create the build with a profile that includes the starting `channel`. For local testing, you can use an [iOS Simulator build](/build-reference/simulators/) or an Android [APK build](/build-reference/apk/).

Reviewer: **[nit]** Platform order: in flowing prose, Expo lists Android before iOS (`style-guide.md#glossary-and-product-names`).

```suggestion
To test channel surfing, install a release build that points to the channel you want to start from. If you use [EAS Build](/build/introduction/), create the build with a profile that includes the starting `channel`. For local testing, you can use an Android [APK build](/build-reference/apk/) or an [iOS Simulator build](/build-reference/simulators/).
```

---
title: Channel surfing
description: Learn how to switch EAS Update channels at runtime.
---

import { BookOpen02Icon } from '@expo/styleguide-icons/outline/BookOpen02Icon';

import { BoxLink } from '~/ui/components/BoxLink';
import { Prerequisites, Requirement } from '~/ui/components/Prerequisites';
import { Step } from '~/ui/components/Step';
import { Terminal } from '~/ui/components/Snippet';

**Channel surfing** lets an installed release build request updates from a different EAS Update channel at runtime.

> In the default EAS Update flow, a build includes an update URL and request headers at build time. The `expo-channel-name` request header determines which channel the build requests updates from. Channel surfing stores a runtime override for that request header on the device.

Reviewer 2: ```suggestion
In the default EAS Update flow, an updates URL and updates channel is fixed at build time. When building on EAS, this channel is set in your **eas.json** build profile. This channel is sent as the `expo-channel-name` request header to the updates URL and determines from which channel your app will receive updates. Channel surfing allows you to override that request header in your JavaScript code, allowing your app to request updates from a different channel.
```

Channel surfing does not change native code or bypass update compatibility rules. The update on the selected channel must still match the installed app's platform and [runtime version](/eas-update/runtime-versions/).

> **warning** Channel surfing with `Updates.setUpdateRequestHeadersOverride()` is available in Expo SDK 54 with `expo-updates` version 0.29.0 and later.

## When to use channel surfing

You can use channel surfing to:

> - Load updates from more than one channel on the same native runtime.

Reviewer 2: The other bullets describe use cases, but this seems to just describe the mechanism itself, so it feels out of place
